nannyml/__init__.py

```diff
@@ -59,7 +59,14 @@
     UnivariateDriftCalculator,
 )
 from .exceptions import ChunkerException, InvalidArgumentsException, MissingMetadataException
-from .io import DatabaseWriter, PickleFileWriter, RawFilesWriter
+from .io import PickleFileWriter, RawFilesWriter
+
+try:
+    from .io.db import DatabaseWriter
+except ImportError:
+    logging.getLogger().warning(
+        "`db` module is not available. Install the `nannyml[db]` extra to use this functionality."
+    )
```

Comment:
Having this here in the `nannyml/__init__.py` file will cause a warning for everyone who uses the `nannyml` package without the db functionality. Probably an improvement over the current behaviour where the dependencies are required, but still doesn't feel great.

We can probably tackle this using a `__getattr__` function on the module, e.g. the second example given in [PEP 562](https://peps.python.org/pep-0562/#rationale). I  imagine this could be something like:

``` python
from importlib import import_module

from .io import PickleFileWriter, RawFilesWriter
...

__all__ = [
    'PickleFileWriter',
    'RawFilesWriter',
    ...
]

# Dict of optional imports mapping object names to module paths
_optional_deps = {
    'DatabaseWriter': '.io.db',
}


def __getattr__(name: str):
    module_path = _optional_deps.get(name)
    if module_path is not None:
        module = import_module(module_path, package=__name__)
        return getattr(module, name)
    raise AttributeError(f'module {__name__!r} has no attribute {name!r}')


def __dir__() -> list[str]:
    return __all__ + list(_optional_deps)
```

With the `ImportError` in place in `.io.db.database_writer` this should result in a good error message only when a user tries to use the `DatabaseWriter` without having the `db` extra installed`.

Unfortunately we'll probably have to do something similar in `nannyml/io/__init__.py`.

Reply:
I agree with the warnings being a bit much - I'll have a look at the PEP and make some changes.

Reply:
I've now pushed some changes for this

Reply:
Amazing, thank you for the update!